Moving camera controller into separate crate in examples

ShadowCurse opened the pull request:
# Objective
Fixes #4408

## Solution

Moved camera controller into `bevy_camera` crate.
Updated current 3d examples to use `bevy_camera`.

DGriffin91 commented:
This is awesome! I think this will be really useful. I like the defaults for up/down and sensitivity.

Would it be better to have the `CameraController` inserted on the perspective camera manually rather than automatically adding it with `check_camera_controller()`?

```rs
commands
    .spawn_bundle(PerspectiveCameraBundle { ..default() })
    .insert(CameraController::default());
```

I don't think this would add much complexity to the examples, and it would both improve flexibility and make it easier for a new user to see how it works.

Also with this configuration `check_camera_controller()` `found_camera` will never be true if a user were to manually assign a CameraController. Not sure if there is a meaningful performance impact.

Would it make sense to have a key that causes it to capture the mouse input?

In Unreal Engine, mouse scroll is used to increase/decrease the moment speed. Would that make sense here?

If we were to more fully duplicate Unreal's input it would actually be right click + drag that rotates the camera. Left click + drag moves the camera forward and backward and rotates left and right. And the middle mouse click + drag moves the camera up/down and left/right.

It may be worthwhile to use these control as a lot of people would be used to them. And it is nice to have the middle click panning. But I also think the current control scheme you have is good, and could make sense as well.

ShadowCurse commented:
Some updates (better late than never):
- Removed `check_camera_controller()` and updated all 3d examples to explicitly add the `CameraController` component. Agree with @DGriffin91 that it better shows how everything works.
- Moved crate from `examples` into `examples_utils` in the root because CI wants to see only executable examples in the `examples` dir.
About upgrading camera movement and so on, I think it can be done in separate PR. 
Furthermore `examples_utils` crate can contain other useful plugins, so upgrading what is inside should be done after this refactoring.

IceSentry commented on examples_utils/src/lib.rs:
```diff
@@ -0,0 +1 @@
+pub mod camera;
```
Since this will show up on docs.rs This needs some top level documentation explaining that this package contains code meant to be reused in examples and not necessarily for user code.

IceSentry commented on examples_utils/src/camera.rs:
```diff
@@ -0,0 +1,141 @@
+use bevy::input::mouse::MouseMotion;
+use bevy::prelude::*;
+
+pub struct CameraControllerPlugin;
```
This needs some documentation. It doesn't have to be much, but just explaining that this is a simple fly camera.

The scene_viewer examples prints the camera controls, but I think this plugin should be in charge of that. Otherwise it will be easy to forget to update it.

IceSentry commented:
@DGriffin91 I'm personally not familiar with unreal. Using the scroll wheel to change the speed sounds really counterintuitive to me and I'm not aware of any camera that does that. 

As someone more familiar with unity, the scroll wheel simply zooms in or out, right click to change the orientation, WASD to move around and middle click to pan.

Blender also only changes the zoom when using the scroll wheel.

Also, middle click panning is not that nice when using a laptop, so having a control scheme that works without a middle click would be nice, but it's not required.

DGriffin91 commented:
@IceSentry I should clarify: in Unreal Engine, it's only when you are holding the right mouse button down that the scroll changes the movement speed. Blender also works the same way when in Walk/Fly mode (normal blender navigation does not use WASD movement, so there is no speed component).

I agree regarding the laptop input. I don't think Unreal has a alternate by default. But I do think we should make sure navigation is good with touchpads as well.

IceSentry commented on examples_utils/src/camera.rs:
```diff
@@ -1,14 +1,17 @@
 impl Plugin for CameraControllerPlugin {
     fn build(&self, app: &mut App) {
-        app.add_system(camera_controller);
+        app.add_system(camera_controller).add_system(print_controls);
```
I don't think this should be a system. It should probably be a startup system

ShadowCurse replied:
Well, my idea was that controls can be changed in `CameraController` when it is attached to the camera. So we need a runtime to print current controls. On the other hand, default controls are fine, and I don't think they will be changed in examples.

IceSentry replied:
Yeah, but adding it in a system like that means it will print it on every frame. Which is just going to spam the console

IceSentry replied:
In that case I'm fine with it

IceSentry commented on examples_utils/src/camera.rs:
```diff
@@ -54,6 +85,16 @@ impl Default for CameraController {
     }
 }
 
+pub fn print_controls(controller: Query<&CameraController>, mut controls_printed: Local<bool>) {
```
Could this use change detection instead?

ShadowCurse replied:
Do you mean to detect if controls have been changed after the initial creation of the `CameraController`? Currently it detects when the camera with the controller is initiated and prints controls of the controller once.

IceSentry replied:
Yes, in your other comment you mentioned wanting to support changing controls. This way if the controls changed during runtime you would see the new controls in the console, but I guess this is fine for now since we don't need that right now.

alice-i-cecile commented on examples_utils/src/lib.rs:
```diff
@@ -0,0 +1,3 @@
+//! This crate contains modules with common functionality used in Bevy engine examples.
+//! They are not intended to be used in user applications.
```
```suggestion
//! They are not intended to be used in user applications, but the code may serve as a useful starting point.
```

alice-i-cecile commented:
Controversial label was added by convention because this adds a new crate; I think this is a pretty straightforward change myself.

ShadowCurse commented on Cargo.toml:
```diff
@@ -122,6 +122,9 @@ bytemuck = "1.7"
 crossbeam-channel = "0.5.0"
 
+# Example crates
+examples_utils = { path = "examples_utils", version = "0.8.0-dev" }
```
The main reason why it is named `example_utils` is so that in the repo it would be near the `examples` dir. But yeah, on crates.io it would look weird. Should I still change it?

ShadowCurse commented:
If we were to convert this to `bevy_camera` then, I think we should introduce some place in the repo where such crates will be. I don’t think placing them in `crates` is a great idea, because everything in `crates` is directly engine related, while `bevy_camera` is more of an addition that is built on top of the engine. But I might be mistaken. In the end, `crates` is so conveniently named.

superdump commented:
Another approach is to not make it into `bevy_camera` and instead just not publish the crate to crates.io, if that is possible. I feel like it is a good idea to have a reasonably good (this would need more work - smoothing and configurability and such) camera controller as part of the project so people can get going quickly and then implement their own when they want to.

mockersf commented:
> I don’t think placing them in `crates` is a great idea, because everything in `crates` is directly engine related, while `bevy_camera` is more of an addition that is built on top of the engine.

I think I disagree? How do you see this distinction between "directly engine related" and "addition built on top"? Nothing would depend on `bevy_camera`? That's already the case of `bevy_audio`. A camera is a relevant part of the engine.

alice-i-cecile commented:
I agree with @mockersf here. I expect we'll accumulate a number of optional leaf node crates over time. cargo is already a great tool to manage that.

ShadowCurse commented:
Separation between “engine related” and “additions built on top” is based on my thoughts about Unreal Engine and the idea that when people think about it, they imagine, not the “engine” itself, but the whole ecosystem surrounding it, that is UEEditor, Blueprint editor and so on. So there is a separation between “the thing that runs everything”(that is the `engine` itself) and “additions”(that is `engine ecosystem`: editor, camera, ui, …). For example, an engine can not run without ECS, but it can function perfectly without a camera controller.

But I digress. I will convert `examples_utils` into `bevy_camera` then. One thing to note though: because `bevy_camera` is not a real “camera”, but just a controller for it, maybe `bevy_camera` should be named `bevy_camera_controller` to make things a little bit more clear?

alice-i-cecile commented:
> One thing to note though: because bevy_camera is not a real “camera”, but just a controller for it, maybe bevy_camera should be named bevy_camera_controller to make things a little bit more clear?

Down the line, I'd like to move all of our camera-related code into there: there's a lot of scattered duplication across the various rendering crates at the moment. So `bevy_camera` gets my vote.

DGriffin91 commented:
@alice-i-cecile If there's going to be a bunch of things in `bevy_camera` (presumably #4572) then should the file with the controller functionality be named more specifically? Like `bevy_camera/camera_controller.rs`?

alice-i-cecile commented:
> @alice-i-cecile If there's going to be a bunch of things in bevy_camera (presumably https://github.com/bevyengine/bevy/pull/4572) then should the file with the controller functionality be named more specifically? Like bevy_camera/camera_controller.rs?

Yes, absolutely :)

ShadowCurse commented:
Hm… I really don’t understand why CI fails. It can’t find a new camera crate, but `bevy_camera` is included in default features and in `bevy_internal` everything seems correct. Locally it compiles and runs without issues.

ShadowCurse commented:
I have checked CI one more time and have found out that bevy is compiled with this command 
```bash
cargo build --no-default-features --features "bevy_dynamic_plugin,bevy_gilrs,bevy_gltf,bevy_winit,render,png,hdr,x11,bevy_ci_testing,trace,trace_chrome,bevy_audio,vorbis"
```
Because of that, the `bevy_camera` feature is not active and examples are failing to compile.
@alice-i-cecile Can you help with this?

alice-i-cecile commented:
> Because of that, the bevy_camera feature is not active and examples are failing to compile.

So, this PR adds `bevy_camera` as a feature to the `Cargo.toml`. The simplest fix would be to just add `bevy_camera` to the list of enabled features for that example-running shell command.

However, if we're planning to move other camera code into here, I think that the correct solution here is to actually add `bevy_camera` [as a subfeature under `render`](https://github.com/bevyengine/bevy/blob/15acd6f45deb3cd21d1f3a9e7d39d135068f09c5/Cargo.toml#L36), which will cause it to be enabled with the existing `render` feature.

ShadowCurse commented:
Alright, I will add it as a `render` subfeature then.

IceSentry commented:
Closed since something similar was merged with https://github.com/bevyengine/bevy/pull/11338
